pkg/protocol/engine/mempool/v1/mempool.go

```diff
@@ -115,20 +115,34 @@ func (m *MemPool[VoteRank]) TransactionMetadata(id iotago.TransactionID) (transa
 	return m.cachedTransactions.Get(id)
 }
 
-// StateMetadata returns the metadata of the state with the given ID.
-func (m *MemPool[VoteRank]) StateMetadata(stateReference iotago.IndexedUTXOReferencer) (state mempool.StateMetadata, err error) {
-	stateRequest, exists := m.cachedStateRequests.Get(stateReference.Ref())
+// OutputStateMetadata returns the metadata of the output state with the given ID.
+func (m *MemPool[VoteRank]) OutputStateMetadata(stateReference *iotago.UTXOInput) (state mempool.OutputStateMetadata, err error) {
+	stateRequest, exists := m.cachedStateRequests.Get(stateReference.StateID())
+
 	if !exists || !stateRequest.WasCompleted() {
-		stateRequest = m.requestStateWithMetadata(stateReference)
+		stateRequest = m.requestState(stateReference)
 	}
 
-	stateRequest.OnSuccess(func(loadedState *StateMetadata) { state = loadedState })
+	stateRequest.OnSuccess(func(loadedState mempool.State) {
+		switch loadedState.Type() {
+		case iotago.InputUTXO:
+			state = loadedState.(mempool.OutputStateMetadata)
+		default:
+			// as we are waiting for this Promise completion, we can assign the outer err
+			err = ierrors.Errorf("invalid state type, only UTXO states can return metadata")
+		}
+	})
 	stateRequest.OnError(func(requestErr error) { err = ierrors.Errorf("failed to request state: %w", requestErr) })
 	stateRequest.WaitComplete()
 
 	return state, err
 }
 
+func (m *MemPool[VoteRank]) PublishCommitmentState(commitment *iotago.Commitment) {
+	stateRequest, _ := m.cachedStateRequests.GetOrCreate(commitment.StateID(), lo.NoVariadic(promise.New[mempool.State]))
+	stateRequest.Resolve(commitment)
```

Comment:
We should not preemtively create promises for commitments. Here we should only resolve existing promises.

If we already have a commitment and a transaction has that on the input side, it should by lazily loaded in `solidifyInputs` logic.

That will make it easier to manage eviction of those.

Reply:
Makes sense, fixed here to only resolve if it was previously requested.

Reply:
InputBlockIssuanceCredit